snapcraft/plugins/python.py

```diff
@@ -0,0 +1,235 @@
+#
+# Copyright (C) 2016 Canonical Ltd
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""The python plugin can be used for python 2 or 3 based parts.
+
+The python plugin can be used for python projects where you would
+want to do:
+
+    - import python modules with a requirements.txt
+    - build a python project that has a setup.py
+    - install packages straight from pip
+
+This plugin uses the common plugin keywords as well as those for "sources".
+For more information check the 'plugins' topic for the former and the
+'sources' topic for the latter.
+
+Additionally, this plugin uses the following plugin-specific keywords:
+
+    - requirements:
+      (string)
+      Path to a requirements.txt file
+    - constraints:
+      (string)
+      Path to a constraints file
+    - process-dependency-links:
+      (bool; default: false)
+      Enable the processing of dependency links.
+    - python-packages:
+      (list)
+      A list of dependencies to get from PyPi
+    - python-version:
+      (string; default: python3)
+      The python version to use. Valid options are: python2 and python3
+"""
+
+import os
+import re
+import subprocess
+from glob import glob
+
+import snapcraft
+from snapcraft import common
+
+
+class PythonPlugin(snapcraft.BasePlugin):
+
+    @classmethod
+    def schema(cls):
+        schema = super().schema()
+        schema['properties']['requirements'] = {
+            'type': 'string',
+        }
+        schema['properties']['constraints'] = {
+            'type': 'string',
+        }
+        schema['properties']['python-packages'] = {
+            'type': 'array',
+            'minitems': 1,
+            'uniqueItems': True,
+            'items': {
+                'type': 'string'
+            },
+            'default': [],
+        }
+        schema['properties']['process-dependency-links'] = {
+            'type': 'boolean',
+            'default': False,
+        }
+        schema['properties']['python-version'] = {
+            'type': 'string',
+            'default': 'python3',
+        }
+        schema.pop('required')
+
+        # Inform Snapcraft of the properties associated with pulling. If these
+        # change in the YAML Snapcraft will consider the pull step dirty.
+        schema['pull-properties'].extend([
+            'requirements',
+            'constraints',
+            'python-packages',
+            'python-version',
+        ])
```

Comment:
What about the build properties? 
I'm not sure how does this work. If the pull properties are changed, and snapcraft build is executed, will it run pull again?